### PR TITLE
drivers sam0: Convert dma to use DEVICE_DT_GET

### PR DESCRIPTION
**Pull request by galak**

### CODE REVIEW

File: drivers/i2c/i2c_sam0.c [hunk 1]
```diff
@@ -247,7 +239,7 @@ static bool i2c_sam0_dma_write_start(const struct device *dev)
 	dma_cfg.channel_direction = MEMORY_TO_PERIPHERAL;
 	dma_cfg.source_data_size = 1;
 	dma_cfg.dest_data_size = 1;
-	dma_cfg.user_data = dev;
+	dma_cfg.user_data = data;
```
Comment (mnkp) [inline drivers/i2c/i2c_sam0.c:242]: Looks suspicious. We haven't modified `i2c_sam0_dma_write_done` callback. It still expects to receive pointer to dev, not pointer to dev->data.

Similar change has been done in `i2c_sam0_dma_read_start`.

File: drivers/spi/spi_sam0.c [hunk 2]
```diff
@@ -706,10 +700,9 @@ static int spi_sam0_init(const struct device *dev)
 	wait_synchronization(regs);
 
 #ifdef CONFIG_SPI_ASYNC
-
-	data->dma = device_get_binding(cfg->dma_dev);
-	data->dev = dev;
```
Comment (mnkp) [inline drivers/spi/spi_sam0.c:711]: It seems that `spi_sam0_dma_rx_done` still expects this to be set.

**Review by pabigot — CHANGES_REQUESTED** [review]
Please address @mnkp comments.

File: drivers/i2c/i2c_sam0.c (hunk 1 above)
Comment (galak) [reply drivers/i2c/i2c_sam0.c:242]: It was hard to tell, I thought user_data wasn't being used at all by i2c_sam0_dma_write_done

Comment (galak) [reply drivers/i2c/i2c_sam0.c]: I see, now.. should be fixed up.

**Review by pabigot — APPROVED** [review]
Seems reasonable.  I don't see a way around the casting-const-away for passing the owning device into the DMA callback.

**Review by dcpleung — APPROVED** [review]

**Review by mnkp — CHANGES_REQUESTED** [review]

**Review by mnkp — APPROVED** [review]

**Review by nandojve — APPROVED** [review]

**Pull request merged**